lib/solarwinds_apm/support/resource_detector.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+require 'net/http'
+require 'uri'
+require 'json'
+
+module SolarWindsAPM
+  # ResourceDetector
+  # Usage:
+  #   require 'opentelemetry/sdk'
+  #   require 'opentelemetry/resource/detector'
+  #   OpenTelemetry::SDK.configure do |c|
+  #     c.resource = SolarWindsAPM::ResourceDetector.detect
+  #   end
+  module ResourceDetector
+    K8S_PODNAME_PATH = '/etc/hostname'
+    K8S_NAMESPACE_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+    K8S_TOKEN_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/token'
+
+    UAMS_CLIENT_PATH = '/opt/solarwinds/uamsclient/var/uamsclientid'
+    UAMS_CLIENT_URL = 'http://127.0.0.1:2113/info/uamsclient'
+    UAMS_CLIENT_ID_FIELD = 'uamsclient_id'
+
+    def self.detect
+      attributes = ::OpenTelemetry::SDK::Resources::Resource.create({})
+      attributes = attributes.merge(detect_uams_client_id)
+      attributes = attributes.merge(detect_k8s_atttributes)
+      attributes.merge(from_upstream_detector)
+    end
+
+    def self.detect_uams_client_id
```

Comment:
Take a look at the [spec](https://github.com/solarwinds-cloud/trace/blob/resource-attributes/docs/Collector-RPC-API.md#uamsclientid) and JS implementation https://github.com/solarwinds/apm-js/blob/main/packages/instrumentations/src/resource-detector-uams.ts, esp. regarding Windows file support and debug logging if possible to help diagnose issues with detection.

Reply:
I thought apm-ruby shouldn't be in windows platform. Is windows file support still necessary?

Reply:
With the upcoming pure-Ruby version, it opens the door to running on Windows. Granted, not a high priority use case but since we're adding this detector might as well make it work.

Reply:
Updated for the path determination based on RUBY_PLATFORM